src/app/clusters/commissioner-control-server/commissioner-control-server.cpp

```diff
@@ -0,0 +1,231 @@
+
+    if (delegate != nullptr)
+    {
+        CHIP_ERROR err = delegate->ReverseCommissionNode(info->params, info->ipAddress, info->port);
+        if (err != CHIP_NO_ERROR)
+        {
+            ChipLogError(Zcl, "ReverseCommissionNode error: %s", err.AsString());
+        }
+    }
+    else
+    {
+        ChipLogError(Zcl, "No delegate available for ReverseCommissionNode");
+    }
+
+    delete info;
+}
+
+} // namespace
+
+namespace chip {
+namespace app {
+namespace Clusters {
+namespace CommissionerControl {
+
+CommissionerControlServer CommissionerControlServer::mInstance;
+
+CommissionerControlServer & CommissionerControlServer::Instance()
+{
+    return mInstance;
+}
+
+CHIP_ERROR CommissionerControlServer::Init(Delegate & delegate)
+{
+    mDelegate = &delegate;
+    return CHIP_NO_ERROR;
+}
+
+Protocols::InteractionModel::Status CommissionerControlServer::GetSupportedDeviceCategoriesValue(
+    EndpointId endpoint, BitMask<SupportedDeviceCategoryBitmap> * supportedDeviceCategories) const
+{
+    Status status = Attributes::SupportedDeviceCategories::Get(endpoint, supportedDeviceCategories);
+    if (status != Status::Success)
+    {
+        ChipLogProgress(Zcl, "CommissionerControl: reading  supportedDeviceCategories, err:0x%x", to_underlying(status));
+    }
+    return status;
+}
+
+Protocols::InteractionModel::Status
+CommissionerControlServer::SetSupportedDeviceCategoriesValue(EndpointId endpoint,
+                                                             const BitMask<SupportedDeviceCategoryBitmap> supportedDeviceCategories)
+{
+    Status status = Status::Success;
+
+    if ((status = Attributes::SupportedDeviceCategories::Set(endpoint, supportedDeviceCategories)) != Status::Success)
+    {
+        ChipLogProgress(Zcl, "CommissionerControl: writing supportedDeviceCategories, err:0x%x", to_underlying(status));
+        return status;
+    }
+
+    return status;
+}
+
+CHIP_ERROR CommissionerControlServer::EmitCommissioningRequestResultEvent(const Events::CommissioningRequestResult::Type & result)
+{
+    EventNumber eventNumber;
+    CHIP_ERROR error = LogEvent(result, kRootEndpointId, eventNumber);
+    if (CHIP_NO_ERROR != error)
+    {
+        ChipLogError(Zcl, "CommissionerControl: Unable to emit CommissioningRequestResult event: %s", error.AsString());
+    }
+
+    return error;
+}
+
+} // namespace CommissionerControl
+} // namespace Clusters
+} // namespace app
+} // namespace chip
+
+bool emberAfCommissionerControlClusterRequestCommissioningApprovalCallback(
+    app::CommandHandler * commandObj, const app::ConcreteCommandPath & commandPath,
+    const Clusters::CommissionerControl::Commands::RequestCommissioningApproval::DecodableType & commandData)
+{
+    CHIP_ERROR err = CHIP_NO_ERROR;
+    Status status  = Status::Success;
+
+    ChipLogProgress(Zcl, "Received command to request commissioning approval");
+
+    auto sourceNodeId = getNodeId(commandObj);
```

Comment:
OK, so this can be a group node ID or some such, right?  Or the PASE made-up ID thing?

The spec requires the event to include "the NodeId of the client which sent the RequestCommissioningApproval command", but that might not be known (e.g. if the command was groupcast).

Reply:
Move this logic into the command handler itself, the Commissioner Control cluster server can only receive RequestCommissioningApproval and CommissionNode command over a CASE session from uni-cast node ID

Reply:
> the Commissioner Control cluster server can only receive RequestCommissioningApproval and CommissionNode command over a CASE session from uni-cast node ID

I don't see the spec saying anything like that.  Where do you see that?

Reply:
Add the check to return kUndefinedNodeId if get the NodeID from a group or PASE session, in the following validation of CommissionNode command, it should fail, since it won't match the NodeId as the RequestCommissioningApproval.

Reply:
Where does the spec say that this is how it should work?

Reply:
In the section of "12.5.7.4. Enabling Fabric Synchronization Between Two Ecosystems"
"After commissioning is complete, the Commissionee Fabric Synchronizing Administrator will now
be accessible on the Fabric administrated by the Commissioner.
Reverse Commissioning
After the Commissioner has finished commissioning the Commissionee, the Commissioner SHOULD
initiate Reverse Commissioning using the Commissioner Control Cluster."

This add the explicit requirement to the commands
https://github.com/CHIP-Specifications/connectedhomeip-spec/pull/10105

But I don't think the PR should be blocked on this additional clarification since we already mandate that the Commissioner SHOULD only initiate Reverse Commissioning using the Commissioner Control Cluster after the Commissioner has finished commissioning the Commissionee.

Reply:
I don't see where any ofthe quoted text says anything about CASE or not.

Reply:
Anyway, https://github.com/CHIP-Specifications/connectedhomeip-spec/pull/10129 now defines the behavior clearly, and we should implement that behavior here in the SDK.

Reply:
The spec PR just merged, update the SDK PR, the function getNodeId already verifies the session type and returns kUndefinedNodeId if the session is not CASE. If getNodeId returns kUndefinedNodeId, we set the status to UNSUPPORTED_ACCESS.